src/v/cloud_storage/cache_service.cc

```diff
@@ -9,49 +9,156 @@
  */
 
 #include "cloud_storage/logger.h"
-#include "random/generators.h"
 #include "utils/gate_guard.h"
+#include "vassert.h"
 #include "vlog.h"
 
 #include <seastar/core/coroutine.hh>
-#include <seastar/core/file.hh>
 #include <seastar/core/fstream.hh>
-#include <seastar/core/future.hh>
-#include <seastar/core/print.hh>
 #include <seastar/core/seastar.hh>
 #include <seastar/core/smp.hh>
-#include <seastar/util/defer.hh>
 
 #include <cloud_storage/cache_service.h>
 
 #include <exception>
-#include <filesystem>
-#include <optional>
-#include <string>
-#include <system_error>
 
 namespace cloud_storage {
 
-cache::cache(std::filesystem::path cache_dir) noexcept
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
   : _cache_dir(std::move(cache_dir))
-  , _cnt(0) {}
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period)
+  , _cnt(0)
+  , _total_cleaned(0) {}
+
+uint64_t cache::get_total_cleaned() { return _total_cleaned; }
+
+bool cache::is_tmp_file(const ss::sstring& filepath) {
+    static const ss::sstring tmp_suffix = ".part";
+    if (filepath.length() <= tmp_suffix.length()) {
+        return false;
+    }
+
+    auto filename_suffix = filepath.substr(
+      filepath.length() - tmp_suffix.length(), filepath.length());
+
+    return !filename_suffix.compare(tmp_suffix);
+}
+
+ss::future<> cache::clean_up_at_start() {
+    gate_guard guard{_gate};
+    auto [unused, candidates_for_deletion] = co_await _walker.walk(
+      _cache_dir.native());
+
+    for (auto& file_item : candidates_for_deletion) {
+        auto filepath_to_remove = file_item.path;
+        vassert(
+          0
+            == filepath_to_remove.substr(0, _cache_dir.native().length())
+                 .compare(_cache_dir.native()),
+          "Tried to clean up {}, which is outside of cache_dir {}.",
+          filepath_to_remove,
+          _cache_dir.native());
+
+        // delete only tmp files that are left from previous RedPanda run
+        if (is_tmp_file(filepath_to_remove)) {
+            try {
+                co_await ss::remove_file(filepath_to_remove);
+                _total_cleaned += file_item.size;
+            } catch (std::exception& e) {
+                vlog(
+                  cst_log.error,
+                  "Cache eviction couldn't delete {}: {}.",
+                  filepath_to_remove,
+                  e.what());
+            }
+        }
+        vlog(
```

Comment:
This probably should be outside the for loop

Reply:
Good catch, thanks!